File: csrc/vframe.cpp
#include "vframe.h"
#include "sym.h"
#include "vstack.h"
#include "vexception.h"
#include "quotestream.h"
#include "term.h"
#include "cmdquote.h"

bool singleassign(); // defined in v.cpp
int VFrame::_idcount = 0;
VFrame::VFrame():_parent(0),_stack(new (collect) VStack()),_id(0) {
    _idcount++;
    _id = _idcount;
}
VFrame::VFrame(VFrame* parent):_parent(parent),_stack(parent->stack()),_id(0) {
    _idcount++;
    _id = _idcount;
}
int VFrame::id() {
    return _id;
}
bool VFrame::hasKey(char* key) {
    QMap::iterator i = _dict.find(key);
    if (i != _dict.end())
        return true;
    return false;
}
Quote* VFrame::lookup(char* key) {
    if (hasKey(key))
        return _dict[key];
    if (_parent)
        return _parent->lookup(key);
    return 0;
}
Quote* VFrame::words() {
    QuoteStream_ nts = new (collect) QuoteStream();
    for(QMap::iterator i = _dict.begin(); i!= _dict.end(); i++) {
        nts->add(new (collect) Term(TSymbol, i->first));
    }
    return new (collect) CmdQuote(nts); 
}
void VFrame::def(char* sym, Quote* q) {
    Char_ s = Sym::lookup(sym);
    if (singleassign() && hasKey(s))
        throw VException("err:symbol_already_bound", new (collect) Term(TSymbol, s), s);
    _dict[s] = q;
}

VFrame* VFrame::parent() {
    return _parent;
}
VFrame* VFrame::child() {
    return new (collect) VFrame(this);
}
VStack* VFrame::stack() {
    return _stack;
}
void VFrame::dump() {
    _stack->dump();
}
void VFrame::reinit() {
    _stack->clear();
}
char* VFrame::to_s() {
    return "<frame>";
}

File: v/VFrame.java
package v;

import java.util.*;

public class VFrame {
    VStack _stack = null;
    HashMap<String, Quote> _dict = new HashMap<String,Quote>();
    VFrame _parent = null;

    public HashMap<String,Quote> dict() {
        return _dict;
    }

    static int _idcount = 0;
    int _id;

    public String id() {
        return "Frame[" + _id + "]";
    }

    public VFrame() {
        _parent = null;
        _stack = new VStack();
        _idcount++;
        _id = _idcount;
    }
    private VFrame(VFrame parent) {
        _parent = parent;
        _stack = parent.stack();
        _idcount++;
        _id = _idcount;
    }
    public Quote lookup(String key) {
        if (_dict.containsKey(key))
            return _dict.get(key);
        if (_parent != null)
            return _parent.lookup(key);
        return null;
    }
    public void def(String sym, Quote q) {
        String s = Sym.lookup(sym);
        if (V.singleassign)
            if (_dict.containsKey(s))
                throw new VException("err:symbol_already_bound", new Term<String>(Type.TString, s),s);
        _dict.put(s,q);
    }
    private VFrame getDefinedScope(String sym) {
        if (_dict.containsKey(sym))
            return this;
        if (_parent == null)
            return null;
        return _parent.getDefinedScope(sym);
    }
    public void set(String sym, Quote q) {
        VFrame dframe = getDefinedScope(sym);
        if (dframe != null)
            dframe.def(sym, q);
        else
            throw new VException("err:symbol_not_defined", new Term<String>(Type.TString, sym),sym);
    }
    public VFrame parent() {
        return _parent;
    }
    public VFrame child() {
        return new VFrame(this);
    }
    public VStack stack() {
        return _stack;
    }
    public void dump() {
        _stack.dump();
    }
    public void reinit() {
        _stack.clear();
    }
}
